### PR TITLE
Reducing generics and indirection in datetime formatters

### PR DESCRIPTION
**Pull request by robertbastian**
~Based on #4863~

### CODE REVIEW

File: components/datetime/src/neo.rs [hunk 1]
```diff
@@ -1963,13 +1968,30 @@ impl<'a> TryWriteable for FormattedNeoDateTime<'a> {
         &self,
         sink: &mut S,
     ) -> Result<Result<(), Self::Error>, fmt::Error> {
-        DateTimeWriter {
-            datetime: &self.datetime,
-            names: self.names,
-            pattern_items: self.pattern.iter_items(),
-            pattern_metadata: self.pattern.metadata(),
+        {
```
Comment (sffc) [inline components/datetime/src/neo.rs:1971]: Question/Suggestion/Issue: It's not immediately clear to me why you aren't using `try_write_pattern` here?

Comment (robertbastian) [reply components/datetime/src/neo.rs:1971]: `try_write_pattern` was generic in the field iterator before. This is the only caller that doesn't use `PatternBorrowed`, but instead constructs a custom iterator. So this monomorphised separately before as well, now it's explicit.

Comment (sffc) [reply components/datetime/src/neo.rs:1971]: ok, if the goal is to make the monomorphization more explicit, I think I'd rather see a `try_write_pattern_from_items` that lives alongside `try_write_pattern`, and `try_write_pattern` can call into it.

Can be a follow-up.

**Review by sffc — APPROVED** [review]

File: components/datetime/src/neo.rs (hunk 1 above)
Comment (robertbastian) [reply components/datetime/src/neo.rs:1971]: `try_write_pattern_from_items` would need to accept a `impl Iterator`, making it very easy to accidentally add another monomorphisation.

Comment (sffc) [reply components/datetime/src/neo.rs:1971]: Sure, but at least you can search for call sites. I would much rather avoid repeating business logic.

**Pull request merged**